File: docs/palindromes-in-peg/scaffold_program.py
"""Run a finite local-tape program on bounded-neighbourhood scaffold cells.

Each invocation of step executes one actual finite instruction. It does not
make an offline program real time. Tape addresses never enter machine labels;
only finite tape symbols and a program counter from the supplied finite table
do. The independent VM enforces backward/self pointer reachability.
"""
from scavm import SELF
from scavm_structs import StackView
from fpp_finite import BLANK


class TapeView:
  def __init__(self, vm, previous, builder, name):
    self.vm, self.builder, self.name = vm, builder, name
    self.focus = BLANK if previous is None else vm.label(previous)[name + ".symbol"]
    self.left = StackView(vm, previous, builder, name + ".l")
    self.right = StackView(vm, previous, builder, name + ".r")

  def read(self):
    return self.focus

  def write(self, symbol):
    self.focus = symbol

  def reset(self):
    """Drop this private persistent tape in bounded work, including its origin."""
    self.focus = BLANK
    self.left.top = self.right.top = None

  def move(self, direction):
    if direction not in (-1, 1):
      raise ValueError("unit tape move required")
    if direction == -1 and self.left.empty():
      raise ValueError("left-end crossing")
    pushed, popped = (self.left, self.right) if direction == 1 else (self.right, self.left)
    slot = self.builder.slot(self.name + ".v")
    self.builder.label[f"{self.name}.v{slot}"] = self.focus
    pushed.push(SELF, slot)
    if popped.empty():
      self.focus = BLANK
    else:
      node, slot = popped.pop2()
      label = self.builder.label if node is SELF else self.vm.label(node)
      self.focus = label[f"{self.name}.v{slot}"]

  def finalize(self):
    self.left.finalize()
    self.right.finalize()
    self.builder.label[self.name + ".symbol"] = self.focus


class ProgramView:
  def __init__(self, vm, previous, builder, program, name=""):
    self.vm, self.builder, self.program = vm, builder, program
    self.prefix = name + "." if name else ""
    self.tapes = [TapeView(vm, previous, builder, f"{self.prefix}t{i}") for i in range(program.ntapes)]
    if previous is None:
      self.pc, self.done = program.start, True
    else:
      label = vm.label(previous)
      self.pc, self.done = int(label[self.prefix + "pc"]), label[self.prefix + "halted"]

  def reset(self):
    # The finite table fixes the number of tapes. No scan or mutation of
    # retained nodes is needed in the persistent scaffold representation.
    for tape in self.tapes:
      tape.reset()
    self.pc, self.done = self.program.start, True

  def start(self, entry=None):
    self.pc = self.program.start if entry is None else entry
    self.done = False

  def step(self):
    if self.done:
      raise ValueError("program is not running")
    row = self.program.code[self.pc]
    op = row[0]
    if op == "halt":
      self.done = True
    elif op == "read":
      self.pc = row[2][self.tapes[row[1]].read()]
    elif op in ("write", "move"):
      getattr(self.tapes[row[1]], op)(row[2])
      self.pc = row[3]
    else:
      raise ValueError("observer-only instruction is not a scaffold operation")

  def finalize(self):
    for tape in self.tapes:
      tape.finalize()
    self.builder.label[self.prefix + "pc"] = str(self.pc)
    self.builder.label[self.prefix + "halted"] = self.done

  def step_read_block(self):
    """Execute one read plus its finite following write/move block.

    Callers must use a table accepted by read_blocks.analyze. The loop bound
    is the fixed table size, never a tape value or an input-dependent length.
    """
    ProgramView.step(self)
    for _ in range(len(self.program.code)):
      if self.done or self.program.code[self.pc][0] in ("read", "halt"):
        return
      ProgramView.step(self)
    raise ValueError("non-reading instruction cycle")

File: scala/pal/src/main/scala/pal/ScaffoldProgram.scala
package pal

import Scavm.{Node, Self, VM}
import ScavmStructs.{Builder, StackView}
import FppFinite.{BLANK, Instruction, Program}

/** Run a finite local-tape program on bounded-neighbourhood scaffold cells.
  *
  * Each invocation of step executes one actual finite instruction. It does not
  * make an offline program real time. Tape addresses never enter machine labels;
  * only finite tape symbols and a program counter from the supplied finite table
  * do. The independent VM enforces backward/self pointer reachability.
  *
  * Python 原典: `scaffold_program.py`（`SCA_ONLINE_CONTROL.md` "Private tape
  * execution"）。Python の `finalize` は `finish`。
  */
object ScaffoldProgram {

  /** A private persistent tape: a focus symbol plus two persistent stacks. Cell
    * contents live in the label `<name>.v<slot>` of the node that created the cell;
    * the stacks hold `(node, slot)` references to those cells.
    */
  final class TapeView(val vm: VM, previous: Option[Node], val builder: Builder, val name: String) {

    var focus: String = previous.fold(BLANK)(p => vm.label(p)(name + ".symbol").asStr)

    val left: StackView = new StackView(vm, previous, builder, name + ".l")
    val right: StackView = new StackView(vm, previous, builder, name + ".r")

    def read(): String = focus

    def write(symbol: String): Unit = {
      focus = symbol
    }

    /** Drop this private persistent tape in bounded work, including its origin. */
    def reset(): Unit = {
      focus = BLANK
      left.top = None
      right.top = None
    }

    /** Unit move: park the focus in a fresh cell of the node being built, then take
      * the focus from the stack on the other side (blank beyond the visited prefix).
      */
    def move(direction: Int): Unit = {
      if (direction != -1 && direction != 1) {
        throw new IllegalArgumentException("unit tape move required")
      }
      if (direction == -1 && left.empty) {
        throw new IllegalArgumentException("left-end crossing")
      }
      val (pushed, popped) = if (direction == 1) { (left, right) } else { (right, left) }
      val slot = builder.slot(name + ".v")
      builder.label(s"$name.v$slot") = focus
      pushed.push(Some(Self), slot)
      if (popped.empty) {
        focus = BLANK
      } else {
        val (cell, vslot) = popped.pop2()
        focus = cellSymbol(cell, vslot)
      }
    }

    /** The symbol parked in cell `(node, vslot)`; the node is [[Self]] for a cell of this step. */
    private def cellSymbol(cell: Option[Scavm.NodeRef], vslot: Int): String = {
      val key = s"$name.v$vslot"
      cell match {
        case Some(Self) => builder.label(key).asStr
        case Some(node: Node) => vm.label(node)(key).asStr
        case None => throw new IllegalStateException(s"tape $name: cell without a creator")
      }
    }

    def finish(): Unit = {
      left.finish()
      right.finish()
      builder.label(name + ".symbol") = focus
    }
  }

  /** A finite program executing on private tapes: the tapes, the program counter
    * and the halted flag, all under the optional prefix `name`.
    *
    * @param name separates simultaneous programs' tape and control fields (`""` = none)
    */
  final class ProgramView(val vm: VM, previous: Option[Node], val builder: Builder, val program: Program, name: String = "") {

    val prefix: String = if (name.nonEmpty) { name + "." } else { "" }

    val tapes: Vector[TapeView] =
      (0 until program.ntapes).map(i => new TapeView(vm, previous, builder, s"${prefix}t$i")).toVector

    var pc: Int = previous.fold(program.start)(p => vm.label(p)(prefix + "pc").asStr.toInt)

    var done: Boolean = previous.fold(true)(p => vm.label(p)(prefix + "halted").asBool)

    /** Drop every private tape and return to the start state, halted.
      *
      * The finite table fixes the number of tapes. No scan or mutation of
      * retained nodes is needed in the persistent scaffold representation.
      */
    def reset(): Unit = {
      for (tape <- tapes) {
        tape.reset()
      }
      pc = program.start
      done = true
    }

    /** Begin running at the program's start state. */
    def start(): Unit = start(program.start)

    /** Begin running at the finite state `entry`. */
    def start(entry: Int): Unit = {
      pc = entry
      done = false
    }

    /** Execute exactly one instruction of the table. */
    def step(): Unit = {
      if (done) {
        throw new IllegalArgumentException("program is not running")
      }
      program.instruction(pc) match {
        case Instruction.Halt =>
          done = true
        case Instruction.Read(tape, choices) =>
          pc = choices(tapes(tape).read())
        case Instruction.Write(tape, symbol, next) =>
          tapes(tape).write(symbol)
          pc = next
        case Instruction.Move(tape, direction, next) =>
          tapes(tape).move(direction)
          pc = next
        case Instruction.Emit(_) =>
          throw new IllegalArgumentException("observer-only instruction is not a scaffold operation")
      }
    }

    def finish(): Unit = {
      for (tape <- tapes) {
        tape.finish()
      }
      builder.label(prefix + "pc") = pc.toString
      builder.label(prefix + "halted") = done
    }

    /** Whether the current instruction ends a read block (`read` or `halt`). */
    private def atReadOrHalt: Boolean = {
      program.instruction(pc) match {
        case Instruction.Read(_, _) | Instruction.Halt => true
        case _ => false
      }
    }

    /** Execute one read plus its finite following write/move block.
      *
      * Callers must use a table accepted by read_blocks.analyze. The loop bound
      * is the fixed table size, never a tape value or an input-dependent length.
      */
    def stepReadBlock(): Unit = {
      step()
      // Python: `for _ in range(len(code)): if done or at read/halt: return; step()`,
      // then raise — so the boundary is tested at most `len(code)` times.
      val settled = program.code.indices.exists { _ =>
        val atBoundary = done || atReadOrHalt
        if (!atBoundary) {
          step()
        }
        atBoundary
      }
      if (!settled) {
        throw new IllegalArgumentException("non-reading instruction cycle")
      }
    }
  }
}
